internal/source/registration.go

```diff
@@ -145,81 +144,104 @@ func (r registration) sourcesForEvent(routes []route, event event.Event) ([]stri
 
 	errs := multierror.New()
 	for _, route := range routes {
-		// event reason
-		match, err := matchRegexForStringIfDefined(route.event.Reason, event.Reason)
+		shouldSend, err := r.shouldSendEventToRoute(route, event)
 		if err != nil {
 			errs = multierror.Append(errs, err)
 			continue
 		}
-		if !match {
-			r.log.Debugf("Ignoring as reason %q doesn't match regex %q", event.Reason, route.event.Reason)
-			continue
-		}
 
-		// event message
-		match, err = matchRegexForStringsIfDefined(route.event.Message, event.Messages)
-		if err != nil {
-			errs = multierror.Append(errs, err)
-			continue
-		}
-		if !match {
-			r.log.Debugf("Ignoring as messages %q don't match regex %q", strings.Join(event.Messages, ";"), route.event.Message)
+		if !shouldSend {
 			continue
 		}
 
-		// resource name
-		match, err = matchRegexForStringIfDefined(route.resourceName, event.Name)
+		out = append(out, route.source)
+	}
+
+	return out, errs.ErrorOrNil()
+}
+
+func (r registration) shouldSendEventToRoute(route route, event event.Event) (bool, error) {
+	log := r.log.WithField("route", route)
+	// event reason
+	if route.event.Reason.AreConstraintsDefined() {
+		match, err := route.event.Reason.IsAllowed(event.Reason)
 		if err != nil {
-			errs = multierror.Append(errs, err)
-			continue
+			return false, err
 		}
 		if !match {
-			r.log.Debugf("Ignoring as resource name %q doesn't match regex %q", event.Name, route.resourceName)
-			continue
+			log.Debugf("Ignoring as reason %q doesn't match constraints %+v", event.Reason, route.event.Reason)
+			return false, nil
 		}
+	}
 
-		// namespace
-		if event.Namespace != "" && !route.namespaces.IsAllowed(event.Namespace) {
-			continue
-		}
+	// event message
+	if route.event.Message.AreConstraintsDefined() {
+		var anyMsgMatches bool
+		var lastErr error
 
-		// annotations
-		if !kvsSatisfiedForMap(route.annotations, event.ObjectMeta.Annotations) {
-			continue
+		eventMsgs := event.Messages
+		if len(eventMsgs) == 0 {
+			// treat no messages as an empty message
+			eventMsgs = []string{""}
 		}
 
-		// labels
-		if !kvsSatisfiedForMap(route.labels, event.ObjectMeta.Labels) {
-			continue
+		for _, msg := range eventMsgs {
+			match, err := route.event.Message.IsAllowed(msg)
+			if err != nil {
+				lastErr = err
+				continue
```

Comment:
What's the purpose of `lastErr`? Why wait for the last error and not break on first error?

Reply:
🤔 Right... It will be always the same error about compiling the same regex. Good catch, will fix that!